File: ruby/hilo.rb
#!/usr/bin/ruby

def hilo
  while true
    puts "\nWelcome to Hi-Lo! The rules are simple: you have 7 tries to guess a number 1-100.\n"\
      "If your guess is too low, I'll say \'Too low\'; if your guess is too high, I'll say \'Too high\'\n"\
      "If you guess correctly, you win!\n"

    answer = rand(1..100)
    
    guess = 0
    for attempt in 1..7
      while true
        print "\nGuess ##{attempt} (1-100): "

        guess = gets.chomp.to_i

        if guess != 0 and guess >= 1 and guess <= 100 
          break
        end
      end

      if guess < answer
        puts "Too low."
      elsif guess > answer
        puts "Too high."
      else
        break
      end
    end

    if guess == answer
      puts "You win!"
    else
      puts "You lost. The correct answer was #{answer}."
    end

    print "Would you like to play again? (y/n):"
    if gets.chomp != "y"
      puts "Thanks for playing!"
      break
    end
  end
end

hilo

File: cxx/hilo.cxx
#include <iostream>
#include <random>

void hilo() {
    std::random_device rd;
    std::mt19937 generator(rd());
    std::uniform_int_distribution<int> distribution(1, 101);

    while (true) {
        std::cout
            << "\nWelcome to Hi-Lo! The rules are simple: you have 7 tries to "
               "guess a number 1-100.\n"
            << "If your guess is too low, I'll say \'Too low\'; if your "
               "guess is too high, I'll say \'Too high\'.\n"
            << "If you guess correctly, you win!\n"
            << std::endl;

        int answer = distribution(generator);

        int guess = 0;
        for (int attempt = 1; attempt <= 7; ++attempt) {
            while (true) {
                std::cout << "Guess #" << attempt << " (1-100): ";

                std::string input;
                std::getline(std::cin, input);

                try {
                    guess = std::stoi(input);
                    if (guess >= 1 && guess <= 100) {
                        break;
                    }
                } catch (...) {
                }
            }

            if (guess < answer) {
                std::cout << "Too low." << std::endl;
            } else if (guess > answer) {
                std::cout << "Too high." << std::endl;
            } else {
                std::cout << "You win!" << std::endl;
                break;
            }
        }

        if (guess != answer) {
            std::cout << "You lost. The correct number was " << answer << "."
                      << std::endl;
        }

        std::cout << "Would you like to play again? (y/n): ";
        std::string choice;
        std::getline(std::cin, choice);
        if (choice != "y") {
            std::cout << "Thanks for playing!" << std::endl;
            break;
        }
    }
}

int main() {
    hilo();

    return 0;
}
